### DISCUSSION TITLE
Прозрачный (transparent) socks используя redsocks

### DISCUSSION THREAD

**Discussion by barl0g** [hufrea/byedpi#71] [General]
Можно ли использовать byedpi в связке с redsocks (pr0cks, transocks и т.д.)? У меня почему-то не получилось, хотя, идее должно работать.
Хотелось бы перенаправить трафик с микротика на локальный сервер с redsocks и byedpi и далее на googlevideo.

Самый удобный вариант был бы. Не надо socks прописывать на клиентах.

**Comment by hufrea** [comment]
> У меня почему-то не получилось, хотя, идее должно работать.

А как вы пытались это сделать?

**Comment by barl0g** [comment]
На домашнем сервере поднял byedpi и redsocks. Настроил на роутере правило
переадресации пакетов на redsocks. Он настроен на работу с byedpi.
Видимо, какой-то мета информации, заголовков не хватает. Так, вообще, может
работать?

пн, 12 авг. 2024 г., 02:13 hufrea ***@***.***>:

**Comment by muzzy124** [comment]
Работает в таком виде (телек с webos работает иногда, телек на тизене - почти никогда):

Поднята машина с redsocks и byedpi (в моем случае alpine c net.ipv4.ip_forward = 1)

Конфиг redsocks.conf:
```
base {
        log_debug = off;
        log_info = on;
        log = stderr;
        daemon = off;
        redirector = iptables;
        redsocks_conn_max = 4096;
}
redsocks {
        local_ip = 0.0.0.0;
        local_port = 12345;
        ip = localhost; // адрес прокси, в данном случае localhost
        port = 1080; // порт прокси
        type = socks5;
}

redudp {
        local_ip = 127.0.0.1; // сетевой адрес машины с редсокс
        local_port = 10053;
        dest_ip = 8.8.8.8; // ну или адрес вашего днс, можно адрес роутера, на котором днс
        dest_port = 53;
        ip = localhost; // адрес прокси, в данном случае localhost
        port = 1080; // порт прокси
        udp_timeout = 30;
        udp_timeout_stream = 180;
}

dnstc {
        local_ip = 127.0.0.1;
        local_port = 5300;
}
```
т.к. пока непонятен вопрос со стабильностью редсокс, перезапускаю сервис такой командой (добавлена в скрипт автозапуска в /etc/local.d)
```
service redsocks start && watch -n 3600 \"service redsocks stop && service redsocks start\"
```

маршруты
```
iptables -t nat -N REDSOCKS
iptables -t nat -A REDSOCKS -d 0.0.0.0/8 -j RETURN
iptables -t nat -A REDSOCKS -d 10.0.0.0/8 -j RETURN
iptables -t nat -A REDSOCKS -d 100.64.0.0/10 -j RETURN
iptables -t nat -A REDSOCKS -d 127.0.0.0/8 -j RETURN
iptables -t nat -A REDSOCKS -d 169.254.0.0/16 -j RETURN
iptables -t nat -A REDSOCKS -d 172.16.0.0/12 -j RETURN
iptables -t nat -A REDSOCKS -d 192.168.0.0/16 -j RETURN
iptables -t nat -A REDSOCKS -d 198.18.0.0/15 -j RETURN
iptables -t nat -A REDSOCKS -d 224.0.0.0/4 -j RETURN
iptables -t nat -A REDSOCKS -d 240.0.0.0/4 -j RETURN

iptables -t nat -A REDSOCKS -p tcp -j REDIRECT --to-ports 12345
iptables -t nat -A REDSOCKS -p udp -j REDIRECT --to-ports 12345

iptables -t nat -A PREROUTING --in-interface eth0 -p tcp -j REDSOCKS
iptables -t nat -A PREROUTING --in-interface eth0 -p udp -j REDSOCKS
```

Варианты запуска byedpi (определяйте опытным путем, а это самое сложное):
```
ciadpi --ip 0.0.0.0 --port 1080 --hosts hosts.txt --disorder 1 --fake 0 --ttl 1 -A torst --tlsrec 1+s --debug 1
ciadpi --ip 0.0.0.0 --port 1080 --hosts hosts.txt --disorder 2 --fake 0 --md5sig --auto=none --debug 1
и т.п.
```

Содержание hosts.txt (субдомены автоматически подхватываются):
```
googlevideo.com
youtube.com
youtu.be
ytimg.com
l.google.com
googleapis.com
gstatic.com
```


Микротик раздает по дхцп (option 3) нужным клиентам шлюз с адресом этой машины c redsocks.

**Reply by barl0g** [reply]
В секцию redudp я ещё добавил local_ip = <proxy_ip> и local_port = 1080. Вроде как quic ютуба не работает без этого. А так, да, вы рабочую схему описали, спасибо.

**Reply by muzzy124** [reply]
поправил конфиг в части redudp, по ошибке не вставил адрес и порт прокси
поменял маршруты на команды

**Reply by muzzy124** [reply]
Короче я устал бороться с тизеном/вебосом и включил ipv6 через брокера (провайдер не умеет в ipv6). На телеках клиенты ютуба в tizenbrew/msx соответственно. Инструкции в гугле... Жду блока через в6 :)

**Reply by barl0g** [reply]
> Короче я устал бороться с тизеном/вебосом и включил ipv6 через брокера (провайдер не умеет в ipv6). На телеках клиенты ютуба в tizenbrew/msx соответственно. Инструкции в гугле... Жду блока через в6 :)

Я пробовал. Глючно тоже как-то работает. Началось с того, что Гугл капчу просил каждый раз (через несколько дней прошло). И частенько другие сайты отваливались. Вот бы сделать так, чтобы только youtube через ipv6 работал. Я не разобрался как это удобно сделать на Микротике.

**Reply by muzzy124** [reply]
по идее парадигма наличия ipv6 уже должна быть стандартной, а то, что что-то не работает, это проблема конкретных приложений или сайтов

конфиги выше обновил до актуальных

**Comment by 5wards** [comment]
Вопрос тот же, только использовал transocks в качестве прозрачного прокси, не работает. 
Помогите плз понять, чего именно не хватает...

**Reply by barl0g** [reply]
Видимо, работает только если redsocks и byedpi находятся на одном сервере. Как я понял, там часть заголовков передаётся через ядро и эта информация теряется при передаче по сети.

**Reply by 5wards** [reply]
на одном и пробовал, такая же связка работает transocks и другой прокси работает отлично, а вот byedpi не хочет

**Reply by muzzy124** [reply]
Выше писал про кейс с отдельным контейнером с byedpi. Сейчас еще и отдельную виртуалку поднял с собранным из сорцов byedpi. Все работает. Машина с редсокс, которая как шлюз выступает, перекидывает весь трафик на машину с сокс5 прокси байдипиай.

**Reply by barl0g** [reply]
У меня что-то redsocks подвисает через несколько часов. Приходится его рестартить. @muzzy124, у вас не подвисает?

**Reply by muzzy124** [reply]
в принципе как-то все нестабильно, то работает, то нет
пробую разные строки, напримре сейчас
ciadpi --ip 0.0.0.0 --port 1080 --hosts hosts.txt --disorder 1 --fake 0 --md5sig --auto=n

**Reply by muzzy124** [reply]
похоже на проблему с редсокс, все как будто виснет, решается рестартом сервиса
в логах что-ли типа

```
1723823582.707064 debug redsocks.c:487 redsplice_write_cb(...) [***:37642->37.9.101.17:443]: backpressure: event_add(relay_read)
1723823582.707125 debug redsocks.c:482 redsplice_write_cb(...) [***:37642->37.9.101.17:443]: backpressure: event_del(relay_read)
1723823582.716794 debug redsocks.c:487 redsplice_write_cb(...) [***:37642->37.9.101.17:443]: backpressure: event_add(relay_read)
```

**Reply by barl0g** [reply]
Вместо redsocks не пробовали что-то?

**Reply by barl0g** [reply]
Возможно, вот это поможет:

`redsocks_conn_max = 4096;`

см. https://github.com/darkk/redsocks/blob/master/redsocks.conf.example

**Reply by muzzy124** [reply]
Спасибо, поставил. Пока тестирую. На всякий случай ещё и перезапуск сервиса редсокс каждые n  минут сделал

**Reply by romkazor** [reply]
Для микрота есть похожая рабочая схема с tun2socks для xray. Поднято 2а контейнера, xray и tun2socks. Трафик от клиента по нужным хостам маркируется и роутится через tun2socks который по socks5 шлет все в xray контейнер - отлично все работает.

С byedpi/spoof-dpi/zapret такая схема почему то не катит, от того же курла получаю `SSL_connect: SSL_ERROR_SYSCALL in connection to youtube.com`. Может кто понимает что нужно подкрутить что бы это взлетело?

**Reply by muzzy124** [reply]
т.е. если менять в схеме xray на byedpi, хотя оба являются сокс5 прокси, перестает работать?

**Reply by romkazor** [reply]
Да, самое что интересное - эта схема не работает с любым из похожих решений как я и писал выше (byedpi/spoof-dpi/zapret), а с xray идеально. При этом если в системе прописать прокси с ip и портом byedpi то все работает, но это не прозрачно.

В дебаге наблюдаю такое полотнище бесконечное:

<details>

-   accept: fd=5
-   new conn: fd=6, addr=74.125.131.190:443
-   host: youtube.com (153)
-   split: pos=0-1 (1), m: DESYNC_DISORDER
-   send: pos=1-517
-   accept: fd=7
-   new conn: fd=8, addr=74.125.131.190:443
-   accept: fd=9
-   host: youtube.com (153)
-   split: pos=0-1 (1), m: DESYNC_DISORDER
-   send: pos=1-517
-   new conn: fd=10, addr=74.125.131.190:443
-   accept: fd=11
-   new conn: fd=12, addr=74.125.131.190:443

</details>

P.S. Кто бы мог подумать, но тут оказалось все проще, микрот нужно было ребутнуть... Почему он не мог нормально работать с новыми добавленными контейнерами для меня большой вопрос.

**Comment by EgorKin** [comment]
А можно Вас попросить реальзовать возможность запуска byedpi в режиме прозрачного прокси?
Дабы не использовать дополнительные утилиты вроде redsocks.

**Reply by Lurker00** [reply]
Очень актуально, особенно потому, что Вы (@hufrea) собираете под MIPS, и этот бинарник прекрасно работает на старых асусовских роутерах! Собрать туда хоть какой-то из имеющихся прозрачных прокси - тот еще квест :(

**Reply by Lurker00** [reply]
А вот даже есть старый форк с добавленным прозрачным режимом: https://github.com/vel21ripn/byedpi/tree/rawtcp

**Reply by barl0g** [reply]
> А вот даже есть старый форк с добавленным прозрачным режимом: https://github.com/vel21ripn/byedpi/tree/rawtcp

(немного оффтоп) В исходниках проскакивает опция --transparent-port, а в описании ничего нет. Пробовали этот режим?

**Reply by Lurker00** [reply]
> проскакивает опция --transparent-port, а в описании ничего нет

Вот это оно самое и есть! Еще не успел попробовать. Много времени заняло исследование, каким тулчейном собирать, собрать этот тулчейн... С утра у меня эта версия работает как SOCKS (т.е. научился собирать), завтра-послезавтра попробую с iptables пробросить трафик.

**Reply by Lurker00** [reply]
Я скрестил прозрачную прокси от @vel21ripn с мастером здесь, и оно прекрасно работает! Причем SmartTube в телевизоре так работает заметно шустрее, чем через явный SOCKS5.

Могу сделать PR, но не понимаю, как, с точки зрения авторских прав, это правильно. Мне проще форкнуть этот репозиторий и подложить уже сделанные правки. Наверное, так и попробую.

**Reply by barl0g** [reply]
> Я скрестил прозрачную прокси от @vel21ripn с мастером здесь, и оно прекрасно работает! Причем SmartTube в телевизоре так работает заметно шустрее, чем через явный SOCKS5.

А без скрещивания что-то не работает?

**Reply by Lurker00** [reply]
Я не пробовал. Тот форк на основе версии 12.0, я посмотрел изменения до 13.1 (текущей) и подумал, что там есть и полезное. Мою версию можно собрать отсюда: https://github.com/Lurker00/byedpi/tree/rawtcp

**Reply by barl0g** [reply]
@hufrea, обратите, пожалуйста, внимание. Возможно, вы решите применить эти изменения в вашем проекте.

**Reply by Lurker00** [reply]
Сделал PR: https://github.com/hufrea/byedpi/pull/114

**Reply by romkazor** [reply]
Спасибо. А более подробно свою схему использования описать можешь?

**Reply by Lurker00** [reply]
Здесь написал: https://4pda.to/forum/index.php?showtopic=491608&view=findpost&p=132158980

**Reply by romkazor** [reply]
Интересует откуда куда и как ходит трафик на byedpi

**Reply by Lurker00** [reply]
> Интересует откуда куда и как ходит трафик на byedpi

byedpi - в маршрутизаторе. Весь трафик из локалки - в него. Дальше - фильтрация через iptables и параметрами --hosts byedpi.

Обновленный набор скриптов тут: https://4pda.to/forum/index.php?showtopic=491608&view=findpost&p=132213916
